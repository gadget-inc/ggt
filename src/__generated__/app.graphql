"""
======================================================
THIS IS A GENERATED FILE! DO NOT EDIT IT MANUALLY!

This is copied from the gadget repository at packages/api/src/__generated__/app.graphql
We manually copy in new versions when we need to change it.
======================================================
"""

schema {
  query: Query
  mutation: Mutation
}

"Instruction for establishing a live connection that is updated once the underlying data changes."
directive @live(
  """Whether the query should be live or not."""
  if: Boolean = true
  """
  Propose a desired throttle interval ot the server in order to receive updates to at most once per "throttle" milliseconds. The server must not accept this value.
  """
  throttle: Int
) on QUERY

"""
A date-time string at UTC, such as 2007-12-03T10:15:30Z, compliant with the `date-time` format outlined in section 5.6 of the RFC 3339 profile of the ISO 8601 standard for representation of dates and times using the Gregorian calendar.
"""
scalar DateTime

"""
One upload target to use for the Direct Upload style of sending files to Gadget
"""
type DirectUploadToken {
  """The token to pass to an action to reference the uploaded file."""
  token: String!
  """The URL to upload a file to."""
  url: String!
}

interface ExecutionError {
  """The Gadget platform error code for this error."""
  code: String!
  """The human facing error message for this error."""
  message: String!
  """
  The stack for any exception that caused the error. Only available for super users.
  """
  stack: String
}

type GadgetAction {
  acceptsInput: Boolean!
  apiIdentifier: String!
  availableInBulk: Boolean!
  bulkApiIdentifier: String
  bulkInvokedByIDOnly: Boolean!
  examples: GadgetActionGraphQLType
  hasAmbiguousIdentifier: Boolean!
  hasCreateOrUpdateEffect: Boolean! @deprecated(reason: "This field will be removed. Use `isCreateOrUpdateAction` instead.")
  hasDeleteRecordEffect: Boolean! @deprecated(reason: "This field will be removed. Use `isDeleteAction` instead.")
  inputFields: [GadgetObjectField!]!
  isCreateOrUpdateAction: Boolean!
  isDeleteAction: Boolean!
  name: String!
  operatesWithRecordIdentity: Boolean!
  possibleTransitions: JSONObject!
  requiresInput: Boolean!
  triggers: [GadgetTrigger!]
}

type GadgetActionGraphQLType {
  bulkOutputGraphQLTypeSDL: String
  exampleBulkGraphQLMutation: String
  exampleBulkGraphQLVariables: JSON
  exampleBulkImperativeInvocation: String
  exampleBulkMutation: String @deprecated(reason: "moved to exampleBulkGraphQLMutation")
  exampleBulkReactHook: String
  exampleGraphQLMutation: String!
  exampleGraphQLVariables: JSON!
  exampleImperativeInvocation: String!
  exampleJSInputs: JSON!
  exampleMutation: String! @deprecated(reason: "moved to exampleGraphQLMutation")
  exampleReactHook: String!
  inputGraphQLTypeSDL: String
  inputTypeScriptInterface: String
  outputGraphQLTypeSDL: String!
  outputTypeScriptInterface: String!
}

type GadgetApplicationMeta {
  allHydrations: JSON!
  """The roles that the entity making this API call has been assigned"""
  assignedRoles: [GadgetRole!]!
  canonicalRenderURL: String!
  developmentGraphQLEndpoint: String!
  developmentRenderURL: String!
  directUploadToken(
    """
    An optional unique value for this upload. Useful for forcing simultaneous requests to not be deduplicated through GraphQL client caching.
    """
    nonce: String
  ): DirectUploadToken
  editURL: String!
  environmentID: GadgetID!
  environmentName: String!
  environmentSlug: String!
  firstModelForExamples: GadgetModel!
  globalActions: [GadgetGlobalAction!]!
  graphQLEndpoint: String!
  hasGlobalActions: Boolean!
  hasLegacyEffectCards: Boolean!
  hasShopifyConnection: Boolean!
  hasSplitEnvironments: Boolean!
  hydrations(modelName: String!): HydrationPlan
  id: GadgetID!
  jsPackageIdentifier: String!
  jsPackageTarballURL: String!
  model(apiIdentifier: String, key: String): GadgetModel
  models: [GadgetModel!]!
  name: String!
  productionGraphQLEndpoint: String!
  productionRenderURL: String!
  referencedHydrations: JSON!
  roles: [GadgetRole!]!
  sessionID: String @deprecated(reason: "The current session is available as the root field `currentSession` on the root Query object, which has the ID as well as other attributes of the session.")
  shopifyConnectionApiVersion: String
  slug: String!
}

type GadgetBelongsToConfig implements GadgetFieldConfigInterface {
  fieldType: GadgetFieldType!
  isConfigured: Boolean!
  isInverseConfigured: Boolean!
  relatedModel: GadgetModel
  relatedModelKey: String
}

type GadgetDateTimeConfig implements GadgetFieldConfigInterface {
  fieldType: GadgetFieldType!
  includeTime: Boolean!
}

type GadgetEnumConfig implements GadgetFieldConfigInterface {
  allowMultiple: Boolean!
  allowOther: Boolean!
  fieldType: GadgetFieldType!
  options: [GadgetEnumOption!]!
}

type GadgetEnumOption {
  color: String!
  name: String!
}

interface GadgetField {
  apiIdentifier: String!
  configuration: GadgetFieldConfigInterface!
  fieldType: GadgetFieldType!
  hasDefault: Boolean!
  name: String!
  required: Boolean!
  requiredArgumentForInput: Boolean!
}

"""The common bits that all field configuration types share"""
interface GadgetFieldConfigInterface {
  fieldType: GadgetFieldType!
}

"""
The type of a given field of a model or other in-transit object within Gadget's type system
"""
enum GadgetFieldType {
  Any
  Array
  BelongsTo
  Boolean
  Code
  Color
  Computed
  DateTime
  Email
  EncryptedString
  Enum
  File
  HasMany
  HasManyThrough
  HasOne
  ID
  JSON
  Money
  Null
  Number
  Object
  Password
  RecordState
  RichText
  RoleAssignments
  String
  URL
  Vector
}

type GadgetFieldUsageExample {
  exampleGraphQLMutation: String!
  exampleGraphQLVariables: JSON!
  exampleImperativeInvocation: String!
  exampleReactHook: String!
}

type GadgetGenericFieldConfig implements GadgetFieldConfigInterface {
  fieldType: GadgetFieldType!
}

type GadgetGlobalAction {
  acceptsInput: Boolean!
  apiIdentifier: String!
  examples: GadgetGlobalActionGraphQLType
  name: String!
  requiresInput: Boolean!
  triggers: [GadgetTrigger!]
}

type GadgetGlobalActionGraphQLType {
  exampleGraphQLMutation: String!
  exampleGraphQLVariables: JSON!
  exampleImperativeInvocation: String!
  exampleJSInputs: JSON!
  exampleMutation: String! @deprecated(reason: "moved to exampleGraphQLMutation")
  exampleReactHook: String!
  inputGraphQLTypeSDL: String
  inputTypeScriptInterface: String
  outputGraphQLTypeSDL: String!
  outputTypeScriptInterface: String!
}

type GadgetHasManyConfig implements GadgetFieldConfigInterface {
  fieldType: GadgetFieldType!
  isConfigured: Boolean!
  isInverseConfigured: Boolean!
  relatedModel: GadgetModel
  relatedModelKey: String
}

"""The ID of a record in Gadget"""
scalar GadgetID

type GadgetModel {
  action(apiIdentifier: String!): GadgetAction
  actions: [GadgetAction!]!
  apiIdentifier: String!
  currentSingletonApiIdentifier: String
  exampleFilterQuery: String!
  exampleFindMostRecentlyCreatedQuery: String!
  exampleFullFindFirstQuery: String!
  exampleFullFindManyQuery: String!
  exampleFullFindOneQuery: String!
  exampleInternalBulkCreateMutation: String!
  exampleInternalCreateMutation: String!
  exampleInternalDeleteManyMutation: String!
  exampleInternalDeleteMutation: String!
  exampleInternalFindFirstQuery: String!
  exampleInternalFindManyQuery: String!
  exampleInternalFindOneQuery: String!
  exampleInternalUpdateMutation: String!
  examplePaginationQuery: String!
  exampleSearchQuery: String!
  exampleSimpleFindManyQuery: String!
  exampleSimpleFindOneQuery: String!
  fields: [GadgetModelField!]!
  filterGraphQLTypeName: String
  filterGraphQLTypeSDL: String
  filterable: Boolean!
  graphQLTypeName: String!
  graphQLTypeSDL: String!
  initialCreatedState: String
  key: String!
  name: String!
  pluralApiIdentifier: String!
  pluralName: String!
  searchable: Boolean!
  sortGraphQLTypeName: String
  sortGraphQLTypeSDL: String
  sortable: Boolean!
  typescriptTypeInterface: String!
  typescriptTypeInterfaceName: String!
}

"""One field of a Gadget model"""
type GadgetModelField implements GadgetField {
  apiIdentifier: String!
  configuration: GadgetFieldConfigInterface!
  examples: GadgetModelFieldExamples!
  fieldType: GadgetFieldType!
  filterable: Boolean!
  hasDefault: Boolean!
  isUniqueField: Boolean!
  name: String!
  required: Boolean!
  requiredArgumentForInput: Boolean!
  sortable: Boolean!
}

type GadgetModelFieldExamples {
  createNestedInParent: GadgetFieldUsageExample
  linkExistingChild: GadgetFieldUsageExample
  linkNewChild: GadgetFieldUsageExample
  linkToExistingParent: GadgetFieldUsageExample
}

"""One field of an action input or other transitory object in Gadget"""
type GadgetObjectField implements GadgetField {
  apiIdentifier: String!
  configuration: GadgetFieldConfigInterface!
  fieldType: GadgetFieldType!
  hasDefault: Boolean!
  name: String!
  required: Boolean!
  requiredArgumentForInput: Boolean!
}

type GadgetObjectFieldConfig implements GadgetFieldConfigInterface {
  fieldType: GadgetFieldType!
  fields: [GadgetModelField!]!
  name: String
}

"""
Represents one of the roles an identity in the system can be entitled to
"""
type GadgetRole {
  key: String!
  name: String!
  order: Int!
  selectable: Boolean!
}

type GadgetTrigger {
  specID: String!
}

"""
Instructions for a client to turn raw transport types (like strings) into useful client side types (like Dates). Unstable and not intended for developer use.
"""
scalar HydrationPlan

type InternalBulkCreateSessionsResult {
  errors: [ExecutionError!]
  sessions: [InternalSessionRecord]
  success: Boolean!
}

type InternalCreateSessionResult {
  errors: [ExecutionError!]
  session: InternalSessionRecord
  success: Boolean!
}

type InternalDeleteManySessionResult {
  errors: [ExecutionError!]
  success: Boolean!
}

type InternalDeleteSessionResult {
  errors: [ExecutionError!]
  session: InternalSessionRecord
  success: Boolean!
}

type InternalMutations {
  """Acquire a backend lock, returning only once the lock has been acquired"""
  acquireLock(
    """Lock name to acquire"""
    lock: String!
  ): LockOperationResult!
  bulkCreateSessions(sessions: [InternalSessionInput]!): InternalBulkCreateSessionsResult
  commitTransaction: String!
  createSession(session: InternalSessionInput): InternalCreateSessionResult
  deleteManySession: InternalDeleteManySessionResult
  deleteSession(id: GadgetID!): InternalDeleteSessionResult
  rollbackTransaction: String!
  startTransaction: String!
  updateSession(id: GadgetID!, session: InternalSessionInput): InternalUpdateSessionResult
}

type InternalQueries {
  """
  Currently open platform transaction details, or null if no transaction is open
  """
  currentTransactionDetails: JSONObject
  listSession(
    """Returns the items in the list that come after the specified cursor."""
    after: String
    """Returns the items in the list that come before the specified cursor."""
    before: String
    """Returns the first n items from the list."""
    first: Int
    """Returns the last n items from the list."""
    last: Int
    """
    A list of field api identifiers to return for each record. If not specified, all fields except for the computed fields on the record are returned.
    """
    select: [String!]
  ): InternalSessionRecordConnection!
  session(
    id: GadgetID!
    """
    A list of field api identifiers to return. If not specified, all fields except for the computed fields on the record are returned.
    """
    select: [String!]
  ): InternalSessionRecord
}

input InternalSessionInput {
  createdAt: DateTime
  id: GadgetID
  """A string list of Gadget platform Role keys to assign to this record"""
  roles: [String!]
  state: RecordState
  stateHistory: RecordState
  updatedAt: DateTime
}

"""
Represents one session result record in internal api calls. Returns a JSON blob of all the record's fields.
"""
scalar InternalSessionRecord

"""A connection to a list of InternalSessionRecord items."""
type InternalSessionRecordConnection {
  """A list of edges."""
  edges: [InternalSessionRecordEdge!]!
  """Information to aid in pagination."""
  pageInfo: PageInfo!
}

"""An edge in a InternalSessionRecord connection."""
type InternalSessionRecordEdge {
  """A cursor for use in pagination"""
  cursor: String!
  """The item at the end of the edge"""
  node: InternalSessionRecord!
}

type InternalUpdateSessionResult {
  errors: [ExecutionError!]
  session: InternalSessionRecord
  success: Boolean!
}

"""
This Error object represents one individual failed validation for a record field. It has a message which is appropriate for display to a user, and lists the apiIdentifier of the field in question. The `apiIdentifier` for the field is not guaranteed to exist on the model.
"""
type InvalidFieldError {
  """The apiIdentifier of the field this error occurred on."""
  apiIdentifier: String!
  """The human facing error message for this error."""
  message: String!
}

"""
This object is returned as an error when a record doesn't pass the defined validations on the model. The validation messages for each of the invalid fields are available via the other fields on this error type.
"""
type InvalidRecordError implements ExecutionError {
  """The Gadget platform error code for this InvalidRecordError."""
  code: String!
  """The human facing error message for this error."""
  message: String!
  """The model of the record which failed validation"""
  model: GadgetModel
  """
  The record which failed validation, if available. Returns all the owned fields of the record -- no sub-selections are required or possible. Only available for super users.
  """
  record: JSONObject
  """The stack for any exception that caused the error"""
  stack: String
  """
  A list of InvalidFieldError objects describing each of the errors encountered on the invalid record.
  """
  validationErrors: [InvalidFieldError!]
  """
  An object mapping field apiIdentifiers to arrays of validation error message strings for that field, as a JSON object. The object may have keys that don't correspond exactly to field apiIdentifiers if added by validations, and the object may have missing keys if no errors were encountered for that field.
  """
  validationErrorsByField: JSONObject
}

"""
The `JSON` scalar type represents JSON values as specified by [ECMA-404](http://www.ecma-international.org/publications/files/ECMA-ST/ECMA-404.pdf).
"""
scalar JSON

"""
The `JSONObject` scalar type represents JSON objects as specified by [ECMA-404](http://www.ecma-international.org/publications/files/ECMA-ST/ECMA-404.pdf).
"""
scalar JSONObject

type LockOperationResult {
  """Any errors encountered during the locking/unlocking operation"""
  errors: [ExecutionError!]
  """Whether the lock operation succeeded"""
  success: Boolean!
}

type Mutation {
  """
  Meta information about the application, like it's name, schema, and other internal details.
  """
  gadgetMeta: GadgetApplicationMeta!
  internal: InternalMutations
}

"""Information about pagination in a connection."""
type PageInfo {
  """When paginating forwards, the cursor to continue."""
  endCursor: String
  """When paginating forwards, are there more items?"""
  hasNextPage: Boolean!
  """When paginating backwards, are there more items?"""
  hasPreviousPage: Boolean!
  """When paginating backwards, the cursor to continue."""
  startCursor: String
}

type Query {
  currentSession: Session
  """
  Meta information about the application, like it's name, schema, and other internal details.
  """
  gadgetMeta: GadgetApplicationMeta!
  internal: InternalQueries
  session(id: GadgetID!): Session
  sessions(
    """Returns the items in the list that come after the specified cursor."""
    after: String
    """Returns the items in the list that come before the specified cursor."""
    before: String
    """Returns the first n items from the list."""
    first: Int
    """Returns the last n items from the list."""
    last: Int
  ): SessionConnection!
}

"""
Represents the state of one record in a Gadget database. Represented as either a string or set of strings nested in objects.
"""
scalar RecordState

"""
A named group of permissions granted to a particular actor in the system. Managed in the Gadget editor.
"""
type Role {
  """
  The stable identifier for this role. Null if the role has since been deleted.
  """
  key: String!
  """The human readable name for this role. Can be changed."""
  name: String!
}

type Session {
  """
  Get all the fields for this record. Useful for not having to list out all the fields you want to retrieve, but slower.
  """
  _all: JSONObject!
  """
  The time at which this record was first created. Set once upon record creation and never changed. Managed by Gadget.
  """
  createdAt: DateTime!
  """
  The globally unique, unchanging identifier for this record. Assigned and managed by Gadget.
  """
  id: GadgetID
  roles: [Role!]
  """
  The time at which this record was last changed. Set each time the record is successfully acted upon by an action. Managed by Gadget.
  """
  updatedAt: DateTime!
}

"""A connection to a list of Session items."""
type SessionConnection {
  """A list of edges."""
  edges: [SessionEdge!]!
  """Information to aid in pagination."""
  pageInfo: PageInfo!
}

"""An edge in a Session connection."""
type SessionEdge {
  """A cursor for use in pagination"""
  cursor: String!
  """The item at the end of the edge"""
  node: Session!
}

"""
This Error object is returned for errors which don't have other specific handling. It has a message which is safe to display to users, but is often technical in nature. It also has a `code` field which is documented in the Gadget API Error Codes docs.
"""
type SimpleError implements ExecutionError {
  """The Gadget platform error code for this error."""
  code: String!
  """The human facing error message for this error."""
  message: String!
  """The stack for any exception that caused the error"""
  stack: String
}