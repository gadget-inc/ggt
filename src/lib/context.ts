import type { Config } from "@oclif/core";
import assert from "assert";
import fs from "fs-extra";
import got, { HTTPError } from "got";
import { isString } from "lodash";
import path from "path";
import { Env } from "./env";
import { ignoreEnoent } from "./fs-utils";

export const GADGET_ENDPOINT = Env.productionLike ? "https://app.gadget.dev" : "https://app.ggt.dev:3000";

class Context {
  /**
   * A reference to oclif's {@linkcode Config}.
   *
   * By default, oclif's {@linkcode Config} is only available as an instance property on a Command, but we want to be able to access it from
   * anywhere. To do this, we created this global variable that references the Config. It is set by the init function in the BaseCommand.
   */
  config!: Config;

  app?: App;

  private _session?: string;

  private _user?: User;

  private _availableApps: App[] = [];

  private _request = got.extend({
    hooks: {
      beforeRequest: [
        (options) => {
          options.headers["user-agent"] = this.config.userAgent;
          if (options.url.origin === GADGET_ENDPOINT && this.session) {
            options.headers["cookie"] = `session=${encodeURIComponent(this.session)};`;
          }
        },
      ],
    },
  });

  get session(): string | undefined {
    if (this._session) return this._session;

    try {
      this._session = fs.readFileSync(path.join(this.config.configDir, "session.txt"), "utf-8");
      return this._session;
    } catch (error) {
      ignoreEnoent(error);
      return undefined;
    }
  }

  set session(value: string | undefined) {
    this.clear();
    this._session = value;
    if (this._session) {
      fs.outputFileSync(path.join(this.config.configDir, "session.txt"), this._session);
    } else {
      fs.removeSync(path.join(this.config.configDir, "session.txt"));
    }
  }

  /**
   * @returns The current user, or undefined if the user is not logged in.
   */
  async getUser(): Promise<User | undefined> {
    if (!this.session) return undefined;
    if (this._user) return this._user;

    try {
      this._user = await this._request(`${GADGET_ENDPOINT}/auth/api/current-user`).json<User>();
      return this._user;
    } catch (error) {
      if (error instanceof HTTPError && error.response.statusCode === 401) {
        this.session = undefined;
        return undefined;
      }
      throw error;
    }
  }

  /**
   * @returns The list of Gadget applications the current user has access to.
   */
  async getAvailableApps(): Promise<App[]> {
    if (!this.session) return [];
    if (this._availableApps.length > 0) return this._availableApps;

    this._availableApps = await this._request(`${GADGET_ENDPOINT}/auth/api/apps`).json<App[]>();
    return this._availableApps;
  }

  async setApp(value: App | string | undefined): Promise<void> {
    if (isString(value)) {
      const app = await this.getAvailableApps().then((apps) => apps.find((app) => app.name == value || app.slug == value));
      assert(app, `attempted to set app to "${value}" but no app with that name or slug was found`);
      this.app = app;
    } else {
      this.app = value;
    }
  }

  clear(): void {
    this._session = undefined;
    this._user = undefined;
    this.app = undefined;
    this._availableApps = [];
  }
}

export const context = new Context();

export interface User {
  email: string;
  name?: string;
}

export interface App {
  id: string | number;
  name: string;
  slug: string;
}
