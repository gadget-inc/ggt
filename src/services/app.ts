import { map } from "lodash";
import { z } from "zod";
import { config } from "./config.js";
import { http, loadCookie } from "./http.js";
import type { User } from "./user.js";

export const App = z.object({
  id: z.union([z.string(), z.number(), z.bigint()]),
  slug: z.string(),
  primaryDomain: z.string(),
  hasSplitEnvironments: z.boolean(),
});

export type App = z.infer<typeof App> & { user: User };

/**
 * @returns The list of Gadget applications the current user has access to.
 */
export const getApps = async (user: User): Promise<App[]> => {
  const cookie = loadCookie();
  if (!cookie) {
    return [];
  }

  const json = await http({
    url: `https://${config.domains.services}/auth/api/apps`,
    headers: { cookie },
    responseType: "json",
    resolveBodyOnly: true,
  });

  return map(z.array(App).parse(json), (app) => ({ ...app, user }));
};
