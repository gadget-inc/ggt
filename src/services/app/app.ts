import assert from "node:assert";
import { z } from "zod";
import type { Context } from "../command/context.js";
import { config } from "../config/config.js";
import { loadCookie } from "../http/auth.js";
import { http } from "../http/http.js";

export const EnvironmentType = Object.freeze({
  Development: "development",
  Production: "production",
  Test: "test",
});

export type EnvironmentType = keyof typeof EnvironmentType;

export const Environment = z.object({
  id: z.union([z.string(), z.number(), z.bigint()]).transform((v) => BigInt(v)),
  name: z.string(),
  type: z.nativeEnum(EnvironmentType),
});

export type Environment = z.infer<typeof Environment>;

// TODO: rename to Application
export const App = z.object({
  id: z.union([z.string(), z.number(), z.bigint()]).transform((v) => BigInt(v)),
  slug: z.string(),
  primaryDomain: z.string(),
  hasSplitEnvironments: z.boolean(),
  multiEnvironmentEnabled: z.boolean(),
  environments: z.array(Environment),
});

export type App = z.infer<typeof App>;

/**
 * Retrieves a list of apps for the given user. If the user is not
 * logged in, an empty array is returned instead.
 *
 * @param ctx - The current context.
 * @returns A promise that resolves to an array of App objects.
 */
// TODO: cache this
export const getApps = async (ctx: Context): Promise<App[]> => {
  const cookie = loadCookie();
  if (!cookie) {
    return [];
  }

  assert(ctx.user, "must get user before getting apps");

  const json = await http({
    context: { ctx },
    url: `https://${config.domains.services}/auth/api/apps`,
    headers: { cookie },
    responseType: "json",
    resolveBodyOnly: true,
  });

  return z.array(App).parse(json);
};
