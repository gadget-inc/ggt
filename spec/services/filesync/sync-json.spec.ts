import fs from "fs-extra";
import { beforeEach, describe, expect, it } from "vitest";
import * as app from "../../../src/services/app/app.js";
import { ArgError } from "../../../src/services/command/arg.js";
import type { Context } from "../../../src/services/command/context.js";
import { Directory } from "../../../src/services/filesync/directory.js";
import { UnknownDirectoryError } from "../../../src/services/filesync/error.js";
import { SyncJson, SyncJsonArgs, type AnySyncJsonState, type SyncJsonStateV05 } from "../../../src/services/filesync/sync-json.js";
import { nockTestApps, testApp, testApp2, testAppWith2Environments } from "../../__support__/app.js";
import { makeContext } from "../../__support__/context.js";
import { expectError } from "../../__support__/error.js";
import { mockOnce } from "../../__support__/mock.js";
import { testDirPath } from "../../__support__/paths.js";
import { loginTestUser } from "../../__support__/user.js";

describe("SyncJson.loadOrInit", () => {
  let localDir: Directory;
  let ctx: Context<SyncJsonArgs>;
  let outputSyncJson: <const State extends AnySyncJsonState>(state: State) => Promise<State>;

  beforeEach(async () => {
    loginTestUser();
    nockTestApps();

    localDir = await Directory.init(testDirPath("local"));
    ctx = makeContext({ parse: SyncJsonArgs, argv: ["dev", `--app=${testApp.slug}`, `--env=${testApp.environments[0]!.name}`] });

    outputSyncJson = async (state) => {
      await fs.outputJSON(localDir.absolute(".gadget/sync.json"), state);
      return state;
    };
  });

  it("loads state from .gadget/sync.json (latest)", async () => {
    const state = await outputSyncJson({
      application: testApp.slug,
      environment: "development",
      environments: {
        development: { filesVersion: "1" },
      },
    });

    const syncJson = await SyncJson.loadOrInit(ctx, { directory: localDir });

    expect(syncJson.state).toEqual(state);
  });

  it("loads state from .gadget/sync.json (v0.4)", async () => {
    await outputSyncJson({
      app: testApp.slug,
      filesVersion: "77",
      mtime: 1658153625236,
    });

    const syncJson = await SyncJson.loadOrInit(ctx, { directory: localDir });

    expect(syncJson.state).toEqual({
      application: testApp.slug,
      environment: "development",
      environments: {
        development: { filesVersion: "77" },
      },
    } as SyncJsonStateV05);
  });

  it("uses default state if .gadget/sync.json does not exist and the directory is empty", async () => {
    const syncJson = await SyncJson.loadOrInit(ctx, { directory: localDir });

    expect(syncJson.state).toEqual({
      application: testApp.slug,
      environment: testApp.environments[0]!.name,
      environments: {
        development: { filesVersion: "0" },
      },
    });
  });

  it(`throws ${UnknownDirectoryError.name} if .gadget/sync.json does not exist and the directory is not empty`, async () => {
    await fs.outputFile(localDir.absolute("foo.js"), "// foo");

    await expect(SyncJson.loadOrInit(ctx, { directory: localDir })).rejects.toThrow(UnknownDirectoryError);
  });

  it(`throws ${UnknownDirectoryError.name} if .gadget/sync.json has invalid json`, async () => {
    await fs.outputFile(localDir.absolute(".gadget/sync.json"), '{"foo":"bar"}');

    await expect(SyncJson.loadOrInit(ctx, { directory: localDir })).rejects.toThrow(UnknownDirectoryError);
  });

  it(`does not throw ${UnknownDirectoryError.name} if .gadget/sync.json has invalid json and --allow-unknown-directory is passed`, async () => {
    await fs.outputFile(localDir.absolute(".gadget/sync.json"), '{"foo":"bar"}');

    ctx = ctx.child({ overwrite: { "--allow-unknown-directory": true } });
    const syncJson = await SyncJson.loadOrInit(ctx, { directory: localDir });

    expect(syncJson.state).toEqual({
      application: testApp.slug,
      environment: testApp.environments[0]!.name,
      environments: {
        development: { filesVersion: "0" },
      },
    });
  });

  it(`throws ${ArgError.name} when --app is passed a slug that does not exist within the user's list of available apps`, async () => {
    ctx = ctx.child({ overwrite: { "--app": "does-not-exist" } });
    const error = await expectError(() => SyncJson.loadOrInit(ctx, { directory: localDir }));

    expect(error).toBeInstanceOf(ArgError);
    expect(error.message).toMatchInlineSnapshot(`
      "Unknown application:

        does-not-exist

      Did you mean one of these?

        • test
        • test2
        • test-with-2-environments
        • test-with-0-environments"
    `);
  });

  it(`throws ${ArgError.name} when --env is passed production`, async () => {
    ctx = ctx.child({ overwrite: { "--env": "production" } });
    const error = await expectError(() => SyncJson.loadOrInit(ctx, { directory: localDir }));

    expect(error).toBeInstanceOf(ArgError);
    expect(error.message).toMatchInlineSnapshot(`
      "Unknown environment:

        production

      Did you mean one of these?

        • development
        • cool-environment-development
        • other-environment-development"
    `);
  });

  it(`throws ${ArgError.name} when --env is passed an environment that is not in the list of valid environments for the app`, async () => {
    ctx = ctx.child({ overwrite: { "--env": "does-not-exist" } });
    const error = await expectError(() => SyncJson.loadOrInit(ctx, { directory: localDir }));

    expect(error).toBeInstanceOf(ArgError);
    expect(error.message).toMatchInlineSnapshot(`
      "Unknown environment:

        does-not-exist

      Did you mean one of these?

        • development
        • cool-environment-development
        • other-environment-development"
    `);
  });

  it(`throws ${ArgError.name} if the user doesn't have any available apps`, async () => {
    mockOnce(app, "getApps", () => []);

    const error = await expectError(() => SyncJson.loadOrInit(ctx, { directory: localDir }));

    expect(error).toBeInstanceOf(ArgError);
    expect(error.message).toMatchInlineSnapshot(`
        "You (test@example.com) don't have have any Gadget applications.

        Visit https://gadget.new to create one!"
    `);
  });

  it(`throws ${ArgError.name} when --app is passed a different slug than the one in .gadget/sync.json`, async () => {
    await outputSyncJson({
      application: testAppWith2Environments.slug,
      environment: testAppWith2Environments.environments[0]!.name,
      environments: {
        development: { filesVersion: "1" },
      },
    });

    const error = await expectError(() => SyncJson.loadOrInit(ctx, { directory: localDir }));

    expect(error).toBeInstanceOf(ArgError);
    expect(error.message).toMatch(/^You were about to sync the following app to the following directory:/);
  });

  it(`does not throw ${ArgError.name} when --app is passed a different app name than the one in .gadget/sync.json and --allow-different-app is passed`, async () => {
    await outputSyncJson({
      application: testApp.slug,
      environment: testApp.environments[0]!.name,
      environments: {
        [testApp.environments[0]!.name]: { filesVersion: "1" },
      },
    });

    ctx = ctx.child({ overwrite: { "--app": testApp2.slug, "--allow-different-app": true } });
    const syncJson = await SyncJson.loadOrInit(ctx, { directory: localDir });

    expect(syncJson.state).toEqual({
      application: testApp2.slug,
      environment: testApp2.environments[0]!.name,
      environments: {
        [testApp2.environments[0]!.name]: { filesVersion: "0" },
      },
    });
  });

  it(`does not throw ${ArgError.name} when --env is passed a different environment than the one in .gadget/sync.json`, async () => {
    await outputSyncJson({
      application: testApp.slug,
      environment: testApp.environments[0]!.name,
      environments: {
        development: { filesVersion: "1" },
      },
    });

    ctx = ctx.child({ overwrite: { "--env": testApp.environments[2]!.name } });
    const syncJson = await SyncJson.loadOrInit(ctx, { directory: localDir });

    expect(syncJson.state).toEqual({
      application: testApp.slug,
      environment: testApp.environments[2]!.name,
      environments: {
        development: { filesVersion: "1" },
        [testApp.environments[2]!.name]: { filesVersion: "0" },
      },
    });
  });

  it("retains environments state when --env flag is passed", async () => {
    const initialState = await outputSyncJson({
      application: testApp.slug,
      environment: testApp.environments[0]!.name,
      environments: {
        [testApp.environments[0]!.name]: { filesVersion: "10" },
        [testApp.environments[2]!.name]: { filesVersion: "14" },
      },
    });

    ctx = ctx.child({ overwrite: { "--env": testApp.environments[3]!.name } });
    const syncJson = await SyncJson.loadOrInit(ctx, { directory: localDir });

    expect(syncJson.state).toEqual({
      ...initialState,
      environment: testApp.environments[3]!.name,
      environments: {
        ...initialState.environments,
        [testApp.environments[3]!.name]: { filesVersion: "0" },
      },
    });
  });
});
