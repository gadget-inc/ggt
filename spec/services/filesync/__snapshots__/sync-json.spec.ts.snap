// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SyncJson.loadOrInit > does not allow --env=production when the command is "add" 1`] = `[Error: You cannot "ggt add" your production environment.]`;

exports[`SyncJson.loadOrInit > does not allow --env=production when the command is "configure" 1`] = `[Error: You cannot "ggt configure" your production environment.]`;

exports[`SyncJson.loadOrInit > does not allow --env=production when the command is "deploy" 1`] = `[Error: You cannot "ggt deploy" your production environment.]`;

exports[`SyncJson.loadOrInit > does not allow --env=production when the command is "dev" 1`] = `[Error: You cannot "ggt dev" your production environment.]`;

exports[`SyncJson.loadOrInit > does not allow --env=production when the command is "list" 1`] = `[Error: You cannot "ggt list" your production environment.]`;

exports[`SyncJson.loadOrInit > does not allow --env=production when the command is "login" 1`] = `[Error: You cannot "ggt login" your production environment.]`;

exports[`SyncJson.loadOrInit > does not allow --env=production when the command is "logout" 1`] = `[Error: You cannot "ggt logout" your production environment.]`;

exports[`SyncJson.loadOrInit > does not allow --env=production when the command is "open" 1`] = `[Error: You cannot "ggt open" your production environment.]`;

exports[`SyncJson.loadOrInit > does not allow --env=production when the command is "push" 1`] = `[Error: You cannot "ggt push" your production environment.]`;

exports[`SyncJson.loadOrInit > does not allow --env=production when the command is "status" 1`] = `[Error: You cannot "ggt status" your production environment.]`;

exports[`SyncJson.loadOrInit > does not allow --env=production when the command is "version" 1`] = `[Error: You cannot "ggt version" your production environment.]`;

exports[`SyncJson.loadOrInit > does not allow --env=production when the command is "whoami" 1`] = `[Error: You cannot "ggt whoami" your production environment.]`;
